packages/@aws-cdk/core/lib/private/refs.ts

```diff
@@ -0,0 +1,276 @@
+  // so construct one in-place.
+  return new Intrinsic({ 'Fn::ImportValue': exportName });
+}
+
+function getCreateExportsScope(stack: Stack) {
+  const exportsName = 'Exports';
+  let stackExports = stack.node.tryFindChild(exportsName) as Construct;
+  if (stackExports === undefined) {
+    stackExports = new Construct(stack, exportsName);
+  }
+
+  return stackExports;
+}
+
+function generateExportName(stackExports: Construct, id: string) {
+  const stack = Stack.of(stackExports);
+  const components = [...stackExports.node.scopes.slice(2).map(c => c.node.id), id];
+  const prefix = stack.stackName ? stack.stackName + ':' : '';
+  const exportName = prefix + makeUniqueId(components);
+  return exportName;
+}
+
+// ------------------------------------------------------------------------------------------------
+// nested stacks
+// ------------------------------------------------------------------------------------------------
+
+/**
+ * Adds a CloudFormation parameter to a nested stack and assigns it with the
+ * value of the reference.
+ */
+function createNestedStackParameter(consumer: Stack, reference: Reference) {
+  // we call "this.resolve" to ensure that tokens do not creep in (for example, if the reference display name includes tokens)
+  const paramId = consumer.resolve(`reference-to-${reference.target.node.uniqueId}.${reference.displayName}`);
+  let param = consumer.node.tryFindChild(paramId) as CfnParameter;
+  if (!param) {
+    param = new CfnParameter(consumer, paramId, { type: 'String' });
+
+    // Ugly little hack until we move NestedStack to this module.
+    if (!('setParameter' in consumer)) {
+      throw new Error('assertion failed: nested stack should have a "setParameter" method');
+    }
+
+    (consumer as any).setParameter(param.logicalId, Token.asString(reference));
+  }
+
+  return param.value as CfnReference;
+}
+
+/**
+ * Adds a CloudFormation output to a nested stack and returns an "Fn::GetAtt"
+ * intrinsic that can be used to reference this output in the parent stack.
+ */
+function createNestedStackOutput(producer: Stack, reference: Reference): CfnReference {
+  const outputId = `${reference.target.node.uniqueId}${reference.displayName}`;
+  let output = producer.node.tryFindChild(outputId) as CfnOutput;
+  if (!output) {
+    output = new CfnOutput(producer, outputId, { value: Token.asString(reference) });
+  }
+
+  if (!producer.nestedStackResource) {
+    throw new Error('assertion failed');
+  }
+
+  return producer.nestedStackResource.getAtt(`Outputs.${output.logicalId}`) as CfnReference;
+}
+
+/**
+ * @returns true if this stack is a direct or indirect parent of the nested
+ * stack `nested`.
+ *
+ * If `child` is not a nested stack, always returns `false` because it can't
+ * have a parent, dah.
+ */
+export function isParent(parent: Stack, child: Stack): boolean {
+  // if "nested" is not a nested stack, then by definition we cannot be its parent
+  if (!child.nestedStackParent) {
```

Comment:
This is required because it is the stopping condition for when the child is *not* a descendent of this parent. Maybe I should reverse the order of the stopping conditions so it will read cleaner:

- If `child.parent === parent` return true
- If `child.parent == null` return false
- otherwise, recurse with `child.parent` as the child

Makes more sense?

Reply:
I also renamed this to `isNested` and reversed the order so it's `isNested(child, parent)`. I hope that's a bit clearer.